components/webcompat/content/browser/webcompat_exceptions_service.cc

```diff
@@ -77,12 +77,10 @@ void WebcompatExceptionsService::LoadWebcompatExceptions(
   base::FilePath txt_file_path =
       install_dir.AppendASCII(WEBCOMPAT_EXCEPTIONS_JSON_FILE_VERSION)
           .AppendASCII(WEBCOMPAT_EXCEPTIONS_JSON_FILE);
-  base::ThreadPool::PostTaskAndReplyWithResult(
+  base::ThreadPool::PostTask(
       FROM_HERE, {base::MayBlock()},
-      base::BindOnce(&brave_component_updater::GetDATFileAsString,
-                     txt_file_path),
-      base::BindOnce(&WebcompatExceptionsService::OnJsonFileDataReady,
-                     weak_factory_.GetWeakPtr()));
+      base::BindOnce(&WebcompatExceptionsService::ReadAndParseJsonRules,
+                     weak_factory_.GetWeakPtr(), txt_file_path));
```

Comment:
> this doesn't look right. weak_factory_.GetWeakPtr() doesn't make sense this way.

Why doesn't it make sense? I think this gives ensures that we run the task with `this` refering to the `WebcompatExceptionsService` instance.

> I don't think you ever need a lock.

I'm happy to do it the way you suggest, but I think we still need the lock in this class for reading and writing `patterns_by_webcompat_type_`, because the callers may call `GetPatterns()` from any thread. Specifically, `HostContentSettingsMap::GetContentSetting` is supposed to be callable from any thread.

Reply:
> Why doesn't it make sense? I think this gives ensures that we run the task with this refering to the WebcompatExceptionsService instance.

you can use weak pointer only on the same thread where `weak_factory_` is invalidated. Right now this doesn't hold, because the weak factory will be invalidated on UI thread, but you're using a weak pointer on a worker thread.

> because the callers may call `GetPatterns()` from any thread

I see. Would be nice to have a comment about this somewhere.

Reply:
Done